contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1047 @@
+    } else {
+      uint256 rampTotalTime = schedule.endTime - schedule.startTime;
+      // slither-disable-next-line timestamp this contract is block time dependent
+      linearAmountAvailable = block.timestamp < schedule.startTime
+        ? 0
+        : (_scheduleTrueTotal(schedule, scheduleId) *
+          (block.timestamp - schedule.startTime)) / rampTotalTime;
+    }
+    return linearAmountAvailable;
+  }
+
+  /* solhint-enable not-rely-on-time */
+
+  /**
+   * @dev Calculates the number of tokens that should be revoked from a given token holder and schedule based on their
+   * proportion of ownership of that schedule's tokens and the total number of tokens being revoked.
+   */
+  function _quantityToRevokePerTokenHolder(
+    uint256 totalQuantityToRevoke,
+    uint256 scheduleId,
+    address account
+  ) private view returns (uint256) {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    uint256 scheduleTrueTotal = _scheduleTrueTotal(schedule, scheduleId);
+    uint256 revocableForAccount;
+    // avoid division by 0
+    if (scheduleTrueTotal == 0) {
+      revocableForAccount = 0;
+    } else {
+      uint256 balanceOfAccount = balanceOf(account, scheduleId);
+      uint256 claimedAmountForAccount = schedule.claimedAmountsByAddress[
+        account
+      ];
+      revocableForAccount =
+        ((claimedAmountForAccount + balanceOfAccount) *
+          (totalQuantityToRevoke)) /
+        scheduleTrueTotal;
+    }
+
+    return revocableForAccount;
+  }
+
+  /**
+   * @dev The total amount of released tokens available at the current block timestamp for the schedule.
+   * Takes the maximum of either the calculated linearly released amount based on the schedule parameters,
+   * or the released amount floor, which is set at the current released amount whenever the balance of a
+   * schedule is decreased through revocation or withdrawal.
+   */
+  function _releasedBalanceOfSingleSchedule(uint256 scheduleId)
+    internal
+    view
+    returns (uint256)
+  {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    return
+      MathUpgradeable.max(
+        _linearReleaseAmountAvailable(scheduleId),
+        schedule.releasedAmountFloor
+      );
+  }
+
+  /** Converts an enumerable set into an array */
+  function _enumerableSetToArray(
+    EnumerableSetUpgradeable.UintSet storage enumerableSet
+  ) internal view returns (uint256[] memory) {
+    uint256[] memory array = new uint256[](enumerableSet.length());
+    for (uint256 i = 0; i < array.length; i++) {
+      array[i] = enumerableSet.at(i);
+    }
+    return array;
+  }
+
+  /**
+   * Reconstructs a schedule's true total based on claimed and unclaimed tokens.
+   *
+   * @dev claiming burns the 1155, so the true total of a schedule has to be reconstructed
+   * from the totalSupply of the token and any claimed amount.
+   */
+  function _scheduleTrueTotal(Schedule storage schedule, uint256 scheduleId)
```

Comment:
seems gratuitous to pass the whole schedule here since it only needs totalClaimedAmount

Reply:
if you do want to pass schedule

Reply:
good idea, and on the other functions... too used to programming with the emphasis on abstraction over tiny efficiencies!